### ISSUE TITLE
Use 2 spaces for unordered sublists, 3 spaces for ordered sublists

### ISSUE THREAD

**Issue by gandalfsaxe** [careteditor/caret#782]
To be compatible with CommonMark + it looks nice/aligned in source code.

[Typora](https://typora.io) already does this. See discussion at DavidAnson/markdownlint/issues/137

**Comment by astoilkov** [comment]
We will discuss this and get back to you. Thanks for the idea.

**Comment by gandalfsaxe** [comment]
For more insightful discussion check neilsustc/vscode-markdown/issues/241

**Comment by erusev** [comment]
It makes sense, I'll see what we can do.

**Comment by seb3point0** [comment]
Perhaps a settings in the preferences file is adequate.

**Comment by gandalfsaxe** [comment]
Correction, both behaviors (one or two spaces after the X.) are actually CommonMark compliant, as noted in the [vscode-markdown discussion](https://github.com/neilsustc/vscode-markdown/issues/241#issuecomment-406989140). However everyone seems to agree that it looks prettier when visually aligned.

If there is wider disagreement on what is prettier, I suppose an option in settings would be reasonable. But I think most people would prefer: 
```
1. one
2. two
   1. three
   2. four
```

to:

```
1. one
2. two
  1. three
  2. four
```

so it's aligned just like:

```
* one
* two
  * three
  * four
```

**Comment by erusev** [comment]
@GandalfSaxe I agree. Thanks for the clarification!

**Comment by borekb** [comment]
As a counter-opinion, I like ~~two~~ four spaces everywhere :)

_Side note_: What I've learned about Markdown over time is is that "computer-processability" is often more important than aesthetics. For example, some of my colleagues like not to write a blank line between a paragraph and a list if those two things form a single piece of information. But then [I remember babelmark](https://babelmark.github.io/?text=Some+paragraph%0A-+followed%0A-+by%0A-+list). I'm not saying that two vs. three spaces are necessarily problematic but to me, it is another opportunity to ask for trouble. My 2c.

**Comment by gandalfsaxe** [comment]
@borekb I can totally understand that sentiment. That's also basically my _modus operandi_ in general. I wasn't aware of this issue until I was getting warnings from VSCode extension markdownlint, and then I start started to notice it and check for it. Turned out Typora was already doing this in the source code, whereas Caret was not.

However I think that forcing people to use code formatters / linters is actually a good thing. I don't mind swallowing a few things that I'd subjectively disagree with myself,  for the sake of everyone's shit being uniformly formatted. Same goes with markdown formatting. My suggest to you: run everyone's markdowncode through [markdownlint](https://github.com/markdownlint/markdownlint) or similarly linter if you're worried about "computer-processability". Because if you're worried about that, you'd need to run it through a formatter / liner anyway exactly because markdown is loosely defined. Point being: making this change is not more opportunity for trouble than there would be anyway; some of your colleagues might use Typora for writing markdown, others Caret or VSCode or Sublime. There's already opportunity for mixing 2 vs. 3 spaces in numbered lists. So the individual tools may as well choose the one they see best fit (as long as it's CommonMark compliant).

**Comment by borekb** [comment]
My point is that even two spaces for the list are [not entirely safe](https://babelmark.github.io/?text=-+list%0A-+with%0A++-+nested%0A++-+items) (MultiMarkdown and Python-Markdown are pretty important implementations). Ok, I know, there's still overwhelming support for two spaces in most Markdown implementations but if trivial examples like this are problematic you can expect a lot more fun with anything fancier.

That's the reality of Markdown, even if I use markdownlint for all documents I write (and I do).

Now, I don't have a strong opinion on what's the right way to go for Caret; just trying to point out possible real-world consequences.

**Comment by gandalfsaxe** [comment]
I don't exactly understand your point. You show a demo of nested _unordered_ lists. What we're talking about here is changing nested _ordered_ lists to having 3 spaces.

Cool tool by the way! It actually drives home the point that 3 spaces are way more conventional than 2 spaces for indented ordered lists:

2 spaces:
https://babelmark.github.io/?text=1.+list%0A2.+with%0A++1.+nested%0A++2.+items

3 spaces:
https://babelmark.github.io/?text=1.+list%0A2.+with%0A+++1.+nested%0A+++2.+items

**Comment by gandalfsaxe** [comment]
Yet another reason for 3 spaces: prettier.io also forces 3 spaces, i.e. 2 spaces or deleted down to 0 spaces, and 4 spaces and made into 3 spaces:

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBGAdAAgDYEsDOMAOlAExYDuuMAFiZg5hplHIXACb2PmbVwBbfCAA0ICAAcYuaMKSgAhgCclECgAVlCOSAUA3CLg6iQAIyUKwAazgwAyhMu4oAc2QwlAVzhjn+OEow6hYuAgrIAGYK2P5iAFb4AB4AQhbWtnYKAnAAMs5wkdGxII5K-krIIGFKVhxqUCYSSs4wAOpGtMgAHAAMYk0Q-q0WEpVNbAF6BWJKcACOnrizwQqh4UhRMT4g-gK47l7b+M4u2HAAip4Q8IVbYjAKpu0cnUik9xa4eK4AwhACYUqUGg0xAnn8ABVHnJNv4AL5woA

**Comment by borekb** [comment]
Oh sorry, of course I meant **four spaces**, I even created an issue for it a long time ago :) https://github.com/careteditor/issues/issues/126. The point I was trying to make here is that it's safer to use _the same amount of spaces_ consistently. Sorry again, to mention two spaces was highly confusing.

Now that I think about it, and after trying to set `tabSize: 2` in Caret which I believe is the default, do ordered lists even work? If I create a nested ordered list in Caret and paste it to GitHub, it won't render as nested, right? That feels almost like a bug and three spaces would be a _fix_.

I still think that four spaces is a better default, as four is the number used in the [CommonMark tutorial](https://commonmark.org/help/tutorial/10-nestedLists.html) and [StackOverflow help](https://stackoverflow.com/editing-help#advanced-lists), and you can google real-world problems with two-spaced nested lists e.g. on BitBucket.

(Another, nitpicky point is how you'd express the `tabSize` lower than four in Caret's settings. Would you set it to `2` even if you knew this actually means three for ordered lists? Would you have two separate settings for ordered and unordered lists? Would you have an ugly but "correct" option like `useAsLowAsPossibleListNesting`? So many questions 😅 )

**Comment by borekb** [comment]
Oh, Caret has actually the `tabSize` option based on my feature request :) Time flies... https://github.com/careteditor/issues/issues/273

**Comment by gandalfsaxe** [comment]
True, while two spaces will work for nested unordered lists in Github, it won't work for nested ordered lists:

* this
* is
  * two
  * spaces
    * in
    * three
    * levels

---

1. this
2. is
  1. two
  2. spaces

---

1. this
2. is
   1. three
   2. spaces

Strong reason to use 3 spaces for ordered nested lists, instead of 2.

Why 3 instead of 4? 1. looks better in source code 2. Seems like the most widely used convention at this point. [Prettier.io](https://prettier.io) enforces 2/3 spaces for unordered/ordered lists, and [markdownlint](https://github.com/DavidAnson/markdownlint)'s default settings likewise give warnings on 4 spaces and expects 2/3 on unordered/ordered.

**Comment by borekb** [comment]
Interesting, I didn't know that markdownlint uses 2 spaces for OLs by default. Anyway, I'm happy enough that Caret gives me `tabSize` and that I can set it to `4` :) BTW, Prettier doesn't enforce 2/3 ([example](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBGAdAAgDYEsDOMAOlAExYDuuMAFiZg5hplHIXACb2PmbVwBbfCAA0ICAAcYuaMKSgAhgCclECgAVlCOSAUA3CLg6iQAIyUKwAazgwAyhMu4oAc2QwlAVzhjn+OEow6hYuAgrIAGYK2P5iAFb4AB4AQhbWtnYKAnAAMs5wkdGxII5K-krIIGFKVhxqUCYSSs4wAOpGtMgAHAAMYk0Q-q0WEpVNbAF6BWJKcACOnrizwQqh4UhRMT4g-gK47l7b+M4u2HAAip4Q8IVbYjAKpu0cnUgALPcWuHiuAMIQAjClSg0GmIE8-gAKo85Jt-ABfeFAA)).
